### PR TITLE
upgrading to nan2 for node4 support

### PR DESCRIPTION
**Pull request by ghost**
This patch is currently ~~NOT~~ ready to merge, but upgrades to nan `^2.1.0` and successfully compiles under node 4. ~~However, the tests currently fail with this message that I haven't been able to track down yet:~~

```
Error: SessionDescription is NULL.
```

This message seems to originate from `third_party/libwebrtc/talk/app/webrtc/peerconnection.cc:505`.

https://github.com/js-platform/node-webrtc/issues/223

### CODE REVIEW

**Comment by caseywebdev** [comment]
Thanks for working on this! I'd love to get this running on node 4.

**Comment by caseywebdev** [comment]
It looks like `local_description` isn't being set on the `PeerConnection`. I've tried to debug where the description gets lost but this is all new territory for me.

File: src/peerconnection.cc [hunk 1]
```diff
@@ -248,51 +248,51 @@ void PeerConnection::OnRenegotiationNeeded() {
...
-  v8::Local<v8::Object> desc = v8::Local<v8::Object>::Cast(args[0]);
-  v8::String::Utf8Value _type(desc->Get(NanNew("type"))->ToString());
-  v8::String::Utf8Value _sdp(desc->Get(NanNew("sdp"))->ToString());
+  PeerConnection* self = Nan::ObjectWrap::Unwrap<PeerConnection>( info.This() );
+  v8::Local<v8::Object> desc = v8::Local<v8::Object>::Cast(info[0]);
+  v8::String::Utf8Value _type(desc->Get(Nan::New("type").ToLocalChecked())->ToString());
+  v8::String::Utf8Value _sdp(desc->Get(Nan::New("sdp").ToLocalChecked())->ToString());
 
   std::string type = *_type;
   std::string sdp = *_sdp;
```
Comment (caseywebdev) [inline src/peerconnection.cc:298]: It looks like `sdp` is incorrectly being set to `true`, and causing the construction of `sdi` below to result in `NULL` (meaning incorrect inputs). `type` is correctly set to `offer`.

Comment (caseywebdev) [inline src/peerconnection.cc:298]: Following the mysterious `true` deeper, it seems the description `peer.createOffer` creates looks like `RTCSessionDescription { type: 'offer', sdp: true }`.

File: src/peerconnection.cc [hunk 2]
```diff
@@ -108,40 +108,40 @@ void PeerConnection::Run(uv_async_t* handle, int status)
...
+      Nan::MakeCallback(pc, callback, 1, argv);
     } else if(PeerConnection::SDP_EVENT & evt.type)
     {
       PeerConnection::SdpEvent* data = static_cast<PeerConnection::SdpEvent*>(evt.data);
-      v8::Local<v8::Function> callback = v8::Local<v8::Function>::Cast(pc->Get(NanNew("onsuccess")));
+      v8::Local<v8::Function> callback = v8::Local<v8::Function>::Cast(pc->Get(Nan::New("onsuccess").ToLocalChecked()));
       v8::Local<v8::Value> argv[1];
-      argv[0] = NanNew(data->desc);
-      NanMakeCallback(pc, callback, 1, argv);
+      argv[0] = Nan::New(&data->desc);
```
Comment (caseywebdev) [inline src/peerconnection.cc]: `data->desc` is the correct SDP string, but `Nan::New(&data->desc)` is `true`. Trying to figure out how to get this working but v8 and Nan is a new frontier.

**Comment by caseywebdev** [comment]
Good news, I have tests passing on node 4 in my branch (https://github.com/caseywebdev/node-webrtc/tree/node4). I've opened a PR on substack's fork (https://github.com/substack/node-webrtc/pull/1) so we can get this all merged into one upstream PR.

**Comment by ghost** [comment]
Now ready to merge. Thanks for the help @caseywebdev!

**Comment by caseywebdev** [comment]
Thanks for doing the hard part!

**Comment by ghost** [comment]
I'm not sure what's up with travis but I can get 0.12 and 4 working locally now.

**Comment by caseywebdev** [comment]
_Probably_ won't make a difference, but apparently these builds are running on "legacy infrastructure".

> This job ran on our legacy infrastructure. Please read our docs on how to upgrade

[The docs](http://docs.travis-ci.com/user/migrating-from-legacy/?utm_source=legacy-notice&utm_medium=banner&utm_campaign=legacy-upgrade) say we just need to add `sudo: false` to `.travis.yml` and move that `sudo apt-get install` into an `addons.apt.packages` array. I figure it couldn't hurt to try.

**Comment by modeswitch** [comment]
If your tests are passing locally, I'll merge this. Travis can be a pain sometimes.

**Comment by rosskukulinski** [comment]
@modeswitch I think we got enabled for OSX builds on travis, but I haven't had time to try it.  I should have some time tomorrow to either A) Manually push node-pre-gyp binaries or B) Have travis auto-build/push.

**Comment by modeswitch** [comment]
Awesome. Let me know if there's anything you need on my end. Once the travis builds are passing I'll push a new version to npm.

**Comment by caseywebdev** [comment]
@modeswitch I've got the builds completing successfully on Linux in #231 if you'd like to take a look.

**Comment by formula1** [comment]
Yay!! ^__^

**Pull request merged**